### PR TITLE
Add better controls for judge connections

### PR DESCRIPTION
**Pull request by Xyene**
Addresses #696.

### CODE REVIEW

File: judge/admin/runtime.py [hunk 1]
```diff
@@ -77,14 +81,29 @@ class JudgeAdmin(VersionAdmin):
...
         (_('Description'), {'fields': ('description',)}),
         (_('Information'), {'fields': ('created', 'online', 'last_ip', 'start_time', 'ping', 'load')}),
         (_('Capabilities'), {'fields': ('runtimes', 'problems')}),
     )
     list_display = ('name', 'online', 'start_time', 'ping', 'load', 'last_ip')
     ordering = ['-online', 'name']
 
+    def get_urls(self):
+        return [url(r'^(\d+)/disconnect/$', self.disconnect_view, name='judge_judge_disconnect'),
+                url(r'^(\d+)/terminate/$', self.terminate_view, name='judge_judge_terminate')] + super(JudgeAdmin, self).get_urls()
```
Comment (quantum5) [inline judge/admin/runtime.py]: Line way too long.

File: judge/bridge/judgehandler.py [hunk 2]
```diff
@@ -108,6 +108,13 @@ def working(self):
     def get_related_submission_data(self, submission):
         return 2, 16384, False, False
 
+    def disconnect(self, force=False):
+        if not force:
```
Comment (quantum5) [inline judge/bridge/judgehandler.py]: `if not` and then `else` is slightly confusing.

File: judge/bridge/judgelist.py [hunk 3]
```diff
@@ -47,6 +47,12 @@ def register(self, judge):
             self.judges.add(judge)
             self._handle_free_judge(judge)
 
+    def disconnect(self, judge_id, force=False):
+        for judge in self.judges:
+            if judge.name == judge_id:
+                judge.disconnect(force=force)
+                return
```
Comment (quantum5) [inline judge/bridge/judgelist.py]: Sometimes, a judge might end up connecting multiple times. I think killing that brain damage would be nice to have, so maybe we shouldn't `return`.

File: judge/models/runtime.py [hunk 4]
```diff
@@ -123,6 +128,12 @@ class Judge(models.Model):
     def __unicode__(self):
         return self.name
 
+    def disconnect(self, force=False):
+        if self.online:
```
Comment (quantum5) [inline judge/models/runtime.py]: A judge might be online in a phantom state. This is known to occur when a judge disconnects due to network failure, and then the judge reinstates the connection. However, the bridge realizes the judge disconnected afterwards, and the judge becomes marked as offline, despite having another connection. We need to significantly rework the bridge to fix this issue, but in the meantime... we probably shouldn't check `self.online`.

**Comment by Xyene** [comment]
cc @quantum5

**Pull request merged**